File: graphrec_core/schemas/datasets.py
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class DatasetSnapshotCreate(BaseModel):
    cutoff_at: datetime | None = None
    description: str | None = None


class DatasetSnapshotResource(BaseModel):
    id: UUID
    tenant_id: UUID
    training_job_id: UUID | None = None
    cutoff_at: datetime
    event_count: int
    product_count: int
    user_count: int
    artifact_uri: str
    checksum: str
    created_at: datetime

    class Config:
        from_attributes = True


class DatasetSnapshotListResponse(BaseModel):
    items: list[DatasetSnapshotResource]


class DatasetUploadResponse(BaseModel):
    accepted_events: int
    accepted_products: int
    dataset_snapshot: DatasetSnapshotResource

File: frontend/src/api/datasets.ts
import { GraphRecApiError } from "./types";

export interface DatasetSnapshotResource {
  id: string;
  tenant_id: string;
  training_job_id: string | null;
  cutoff_at: string;
  event_count: number;
  product_count: number;
  user_count: number;
  artifact_uri: string;
  checksum: string;
  created_at: string;
}

export interface DatasetSnapshotListResponse {
  items: DatasetSnapshotResource[];
}

export interface DatasetUploadResponse {
  accepted_events: number;
  accepted_products: number;
  dataset_snapshot: DatasetSnapshotResource;
}

async function parseResponseJson(res: Response): Promise<any> {
  try {
    return await res.json();
  } catch {
    return {
      code: "http_error",
      message: res.statusText || `Server returned HTTP status ${res.status}`,
    };
  }
}

export async function uploadDatasetFile(
  token: string,
  file: File,
): Promise<DatasetUploadResponse> {
  const formData = new FormData();
  formData.append("file", file);

  const res = await fetch("/v1/datasets/upload", {
    method: "POST",
    headers: { Authorization: `Bearer ${token}` },
    body: formData,
  });

  const body = await parseResponseJson(res);
  if (!res.ok) throw new GraphRecApiError(res.status, body);
  return body as DatasetUploadResponse;
}

export async function createDatasetSnapshot(
  token: string,
  cutoffAt?: string,
): Promise<DatasetSnapshotResource> {
  const res = await fetch("/v1/datasets/snapshots", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ cutoff_at: cutoffAt || null }),
  });

  const body = await parseResponseJson(res);
  if (!res.ok) throw new GraphRecApiError(res.status, body);
  return body as DatasetSnapshotResource;
}

export async function listDatasetSnapshots(token: string): Promise<DatasetSnapshotListResponse> {
  const res = await fetch("/v1/datasets/snapshots", {
    headers: { Authorization: `Bearer ${token}` },
  });

  const body = await parseResponseJson(res);
  if (!res.ok) throw new GraphRecApiError(res.status, body);
  return body as DatasetSnapshotListResponse;
}
